### PR TITLE
Implement support for optimizing away the dynamic registrar.

### PR DESCRIPTION
**Pull request by rolfbjarne**
* Add a new property (Runtime.DynamicRegistrationSupported) that indicates at
  runtime whether dynamic registration is available.

* Check this new property whenever we need dynamic registration.

* Add an optimization to mtouch/mmp that detects whether dynamic registration
  is required, and if not, changes Runtime.DynamicRegistrationSupported to
  return false.

* Add an optimization to mtouch/mmp that inlines
  Runtime.DynamicRegistrationSupported as a constant value.

The end result is that the linker will link away the dynamic registrar if
mtouch/mmp detects that it's not needed.

Benchmark
---------

I've compared the size of entire apps built for device:

|test                          | Before |  After |   Diff |     % |
|:-----------------------------|-------:|-------:|-------:|------:|
|[monotouch-test/Release][1]   | 95.7mb | 95.0mb | -680kb | -0.7% |
|[link sdk/Release][2]         | 21.2mb | 20.9mb | -245kb | -1.2% |
|[minimalistic app/Release][3] | 4.58mb | 4.32mb | -259kb | -5.7% |

[1]: https://gist.github.com/rolfbjarne/3871e36d2de8db1a8eee1d9f9276d3d2#monotouch-test-release-wall-optimizations-enabled
[2]: https://gist.github.com/rolfbjarne/3871e36d2de8db1a8eee1d9f9276d3d2#link-sdk
[3]: https://gist.github.com/rolfbjarne/3871e36d2de8db1a8eee1d9f9276d3d2#minimalistic-app

### CODE REVIEW

File: docs/website/mmp-errors.md [hunk 1]
```diff
@@ -236,6 +236,12 @@ Please file an [issue](https://github.com/xamarin/xamarin-macios/issues/new)
 along with a complete build log so that we can investigate what went wrong and
 possibly enable more scenarios in the future.
 
+### <a name="MM2107"/>MM2107: It's not safe to remove the dynamic registrar because {reasons}
+
+The linker reports this warning when the developer requests removal of the
+dynamic registrar (by passing `--optimize:remove-dynamic-registrar` to
+mmp), but the linker determines that it's not safe to do so.
+
```
Comment (spouliot) [inline docs/website/mmp-errors.md]: missing action ;-) e.g.

The default optimization option was overridden in your project. Remove it to get rid of the warning at build time.
Note: By default this option will be automatically enabled whenever it's possible/safe to apply.

File: docs/website/mtouch-errors.md [hunk 2]
```diff
@@ -1311,6 +1311,12 @@ Please file an [issue](https://github.com/xamarin/xamarin-macios/issues/new)
 along with a complete build log so that we can investigate what went wrong and
 possibly enable more scenarios in the future.
 
+### <a name="MT2107"/>MT2107: It's not safe to remove the dynamic registrar because {reasons}
+
+The linker reports this warning when the developer requests removal of the
+dynamic registrar (by passing `--optimize:remove-dynamic-registrar` to
+mmp), but the linker determines that it's not safe to do so.
+
```
Comment (spouliot) [inline docs/website/mtouch-errors.md]: same

File: docs/website/optimizations.md [hunk 3]
```diff
@@ -380,3 +380,71 @@ Xamarin.Mac supports loading assemblies dynamically, and those assemblies
 might not have been known at build time (and thus not optimized).
 
 The default behavior can be overridden by passing `--optimize=-register-protocols` to mtouch/mmp.
+
+Remove the dynamic registrar
+----------------------------
+
+Both Xamarin.iOS and Xamarin.Mac include support for [registering managed
+types][1] with the Objective-C runtime. It can either be done at build time or
```
Comment (spouliot) [inline docs/website/optimizations.md]: ^ I don't think that syntax works with mark-down (on the website), you have a inline the URL within a `()`

File: tests/common/BundlerTest.cs [hunk 4]
```diff
@@ -106,6 +106,63 @@ static void Main ()
 			}
 		}
 
+#if __MACOS__
+		// XM doesn't support removing the dynamic registrar yet.
+		//[TestCase (Profile.macOSMobile)]
```
Comment (chamons) [inline tests/common/BundlerTest.cs]: Is this still true? There is text above about XM not being able to guess but you can pass the command line arg to force it.

File: tools/linker/CoreTypeMapStep.cs [hunk 5]
```diff
@@ -25,19 +26,142 @@ namespace MonoTouch.Tuner {
...
+			if (LinkContext.App.Optimizations.RemoveDynamicRegistrar != false)
+				dynamic_registration_support_required |= RequiresDynamicRegistrar (assembly, LinkContext.App.Optimizations.RemoveDynamicRegistrar == true);
+
+			base.ProcessAssembly (assembly);
+		}
+
+		// If certain conditions are met, we can optimize away the code for the dynamic registrar.
+		bool RequiresDynamicRegistrar (AssemblyDefinition assembly, bool warnIfRequired)
+		{
+			// Disable removing the dynamic registrar for XM/Classic to simplify the code a little bit.
```
Comment (chamons) [inline tools/linker/CoreTypeMapStep.cs]: XM classic shouldn't be building anymore so technically not needed, but still nice to note.

Comment (rolfbjarne) [reply tools/linker/CoreTypeMapStep.cs]: mmp still supports building Classic apps

File: tests/common/BundlerTest.cs (hunk 4 above)
Comment (rolfbjarne) [reply tests/common/BundlerTest.cs]: Yes, this is still true, I found problems with it for XM quite late and didn't have to fix them before d15-7.

File: tools/linker/CoreTypeMapStep.cs (hunk 5 above)
Comment (chamons) [reply tools/linker/CoreTypeMapStep.cs]: This is true. duh

**Comment by spouliot** [comment]
build (tests) failed due to lack of disk space
monitoring results from wrench instead

**Comment by rolfbjarne** [comment]
Wrench results: https://wrench.internalx.com/Wrench/ViewLane.aspx?lane_id=5746&host_id=392&revision_id=929081

All test failures are unrelated:

* linkall/Mac/Debug: [linkall/Mac/Debug build failure](https://github.com/xamarin/maccore/issues/642)
* Mono.Security/Mac/Modern: [Test failure on bots: Mono.Security/Mac/Modern: MonoTests.Mono.Security.Authenticode.AuthenticodeDeformatterTest.VerifySignedAssembly](https://github.com/xamarin/xamarin-macios/issues/3207)
* MSBuild/iOS: [[XI]WatchOS apps fail to build with error "Undefined symbols for architecture armv7k" using XI with mono-2017-12 support](https://github.com/mono/mono/issues/6948)
* MTouch tests: [MTouch tests: Xamarin.MTouch.RebuildTest_WithExtensions failure](https://github.com/xamarin/maccore/issues/641)
* xammac/Release: [[jenkins] XM/introspection random crashes](https://github.com/xamarin/maccore/issues/568)

**Pull request merged**